Add config flow for Bose SoundTouch

kroimon opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->
Port configuration is no longer possible as all SoundTouch devices use port 8090 for the API anyway. This should therefore affect no one because there was no reason to set this to anything else than 8090 in the past.

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This adds a configuration flow for Bose SoundTouch devices.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/23241

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [X] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [X] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [X] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [X] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

kroimon committed:
Add config flow for Bose SoundTouch

epenet commented:
Hi, is there a way to split up this PR to reduce the size of the review? In particular, all the work that you can do ahead of the config flow would help.

As an example:
- preliminary PR for codeowner
- preliminary PR for refactoring the tests (though I am not sure if moving to requests_mock would be accepted)
- preliminary PR to move the services to `__init__.py` (if that's possible)
- config-flow
- etc.

kroimon commented:
Thanks for your feedback, @epenet!
I was aware that this PR is a bit on the big side, but didn't really know how I could split it up. I started working on the config flow and then realized that the tests were kind of messy and I didn't just want to add more messy tests for the config flow changes, so my changes got a bit out of hand ;)

I created a separate, smaller PR for the tests to merged first: #72984 

I can also split out the code owner change later if any of my changes are merged and I actually "own" some of the code.

kroimon commented:
I rebased this PR to remove changes from the already merged PR #72984, so this is now ready for review again.

The failing tests are unrelated to this PR (current dev was broken by #64927).

balloob commented on homeassistant/components/soundtouch/__init__.py:
```diff
@@ -1 +1,142 @@
+        vol.Required("slaves"): cv.entity_ids,
+    }
+)
+SERVICE_ADD_ZONE_SCHEMA = vol.Schema(
```
Media player has built-in support for join and unjoin. we should use that.

kroimon replied:
I just ported this over from the previous implementation because I didn't want to get this PR even more complicated.
Join and Unjoin services are on my to-do list for a follow-up PR.

balloob commented on homeassistant/components/soundtouch/__init__.py:
```diff
@@ -1 +1,142 @@
+        elif service.service == SERVICE_ADD_ZONE_SLAVE:
+            await hass.async_add_executor_job(master.add_zone_slave, slaves)
+
+    hass.services.async_register(
```
So for any of the services that are going to be target media players, register it on the `entity_platform` [like this](https://github.com/home-assistant/core/blob/dev/homeassistant/components/channels/media_player.py#L57-L63)

balloob commented on homeassistant/components/soundtouch/config_flow.py:
```diff
@@ -0,0 +1,107 @@
+            title=self.name or self.host,
+            data={
+                CONF_HOST: self.host,
+                CONF_NAME: self.name,
```
Why store name? The entry already has a name.

balloob commented on homeassistant/components/soundtouch/config_flow.py:
```diff
@@ -0,0 +1,107 @@
+            data_schema=vol.Schema(
+                {
+                    vol.Required(CONF_HOST): cv.string,
+                    vol.Optional(CONF_NAME, default=""): cv.string,
```
Why do we need to ask for the name?

kroimon replied:
I added the name because it was a property in the yaml configuration, but as this can be overridden for the media player entity via the GUI later, it should indeed be unnecessary. I will remove it.

balloob commented on homeassistant/components/soundtouch/media_player.py:
```diff
@@ -50,137 +49,58 @@
-        vol.Optional(CONF_PORT, default=DEFAULT_PORT): cv.port,
-    }
+PLATFORM_SCHEMA = vol.All(
+    cv.deprecated(CONF_PORT),
```
Everything is deprecated, not just port.

kroimon replied:
I went with what I found in other components: logging the migration notice any deprecating the whole config section in the next version? Should I completely deprecate it now?

balloob replied:
you already print a deprecation warning inside `async_setup_platform` so this single key deprecation is not necessary.

balloob commented on homeassistant/components/soundtouch/media_player.py:
```diff
@@ -50,137 +49,58 @@
+
+    async_add_entities([media_player], True)
+
+    hass.data[DOMAIN][entry.entry_id].media_player = media_player
```
We shouldn't let other parts of the integration directly call methods on this media player. Once you switch to registering services via `entity_platform` you don't need to anyway 👍

kroimon replied:
Correct. I planned on removing this once the join/unjoin services are in place

balloob commented on homeassistant/components/soundtouch/media_player.py:
```diff
@@ -197,28 +117,32 @@ class SoundTouchDevice(MediaPlayerEntity):
+        self._device = device
+
+        self._attr_unique_id = self._device.config.device_id
+        self._attr_name = name or self._device.config.name
```
drop `name or`. Just stick to device name.

balloob commented on homeassistant/components/soundtouch/strings.json:
```diff
@@ -0,0 +1,24 @@
+  "config": {
+    "step": {
+      "user": {
+        "title": "Setup a new Bose SoundTouch device",
+        "description": "Enter the device hostname or IP address of the Bose SoundTouch you want to add. Optionally, you can also specify the name to use.",
```
Title is automatically set to manifest title if omitted, so not needed here.
Description also doesn't add much info. So let's drop it.
```suggestion
```

balloob commented on homeassistant/components/soundtouch/strings.json:
```diff
@@ -0,0 +1,24 @@
+        "title": "Setup a new Bose SoundTouch device",
+        "description": "Enter the device hostname or IP address of the Bose SoundTouch you want to add. Optionally, you can also specify the name to use.",
+        "data": {
+          "host": "[%key:common::config_flow::data::host%]",
+          "name": "[%key:common::config_flow::data::name%]"
```
Drop name.
```suggestion
          "host": "[%key:common::config_flow::data::host%]"
```

balloob reviewed (commented):
Off to a great start!

kroimon commented on homeassistant/components/soundtouch/__init__.py:
```diff
@@ -1 +1,142 @@
+
+    hass.services.async_register(
+        DOMAIN,
+        SERVICE_PLAY_EVERYWHERE,
```
This was already existing and can be deprecated in a future PR, I guess?

kroimon committed:
Use device name instead of user-defined name

kroimon committed:
Remove depreciation warning for port configuration

kroimon commented:
@balloob Thanks a lot for the extensive review.
I removed the name from the configuration and changed it to always use the name reported by the device itself by default. The extra deprecation warning for the port has also been removed.
I will keep the previously existing services for now and try to replace them with the new media player platform services in the future.

I don't really understand why the CI tests are failing as they complete successfully when running them locally. Am I missing any important clean-ups or teardowns in the tests?

balloob committed:
Mock async_setup_entry

balloob commented on tests/components/soundtouch/test_config_flow.py:
```diff
@@ -0,0 +1,113 @@
+    assert result.get("step_id") == "user"
+    assert "flow_id" in result
+
+    with patch(
```
The async_setup_entry were not patched, so your tests were trying to make requests to soundtouch devices.

balloob replied:
Fixed in e42bef8140eb83a0863562fc26e2aabc47a89c2a

kroimon replied:
Awesome, thank you!

MartinHjelmare commented on homeassistant/components/soundtouch/__init__.py:
```diff
@@ -1 +1,142 @@
+            ]
+
+        master = next(
+            iter(
```
This can just be a generator expression.

```py
gen_expr = (item for item in items if condition)
```

MartinHjelmare commented on homeassistant/components/soundtouch/__init__.py:
```diff
@@ -1 +1,142 @@
+    )
+    hass.services.async_register(
+        DOMAIN,
+        SERVICE_CREATE_ZONE,
```
Is creating a zone similar to grouping sound devices to play the same thing? If so, we have a standard media_player interface for that.

https://developers.home-assistant.io/docs/core/entity/media-player#grouping-player-entities-together

MartinHjelmare commented on homeassistant/components/soundtouch/__init__.py:
```diff
@@ -1 +1,142 @@
+PLATFORMS = [Platform.MEDIA_PLAYER]
+
+
+class SoundTouchData:
```
This could be a dataclass instead.

https://docs.python.org/3/library/dataclasses.html

MartinHjelmare commented on homeassistant/components/soundtouch/media_player.py:
```diff
@@ -232,17 +155,15 @@ def volume_level(self):
             return STATE_OFF
 
+        if self._status.source == "INVALID_SOURCE":
+            return STATE_UNKNOWN
```
We should return `None` as state to represent unknown state. It will be converted by the base entity class.

MartinHjelmare commented on tests/components/soundtouch/test_config_flow.py:
```diff
@@ -0,0 +1,113 @@
+    assert result["errors"] == {"base": "cannot_connect"}
+
+
+async def test_zeroconf_flow_create_entry(
```
Please add a test for already configured device for user and zeroconf flows. Add a MockConfigEntry and then start the flow.

MartinHjelmare reviewed (commented):
Please address the comments in a new PR. Thanks!
